kt-review.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
+import random
+import sys
+
+def main():
+    """Conduct a knight's tour using a brute force method, representing the
+       board as a dict.
+    """
+
+    if not valid_input():
+        sys.exit("usage: knightstour4.py rows columns attempts")
+
+    rows = int(sys.argv[1])
+    cols = int(sys.argv[2])
+    attempts = int(sys.argv[3])
+
+    for attempt in range(attempts):
+        board = create_board(rows, cols)
+        knight_pos = [0, 0]
+
+        for j in range(rows * cols):
+            board[tuple(knight_pos)] = j
+            possible_moves = generate_possible_moves(knight_pos, board, rows, cols)
+
+            if possible_moves == ():
+                break
+
+            move = random.choice(possible_moves)
+
+            knight_pos[0] += move[0]
+            knight_pos[1] += move[1]
+
+        # at this point, all possible moves have been performed, and the knight's
+        # tour is either complete or incomplete
+        if None not in board.values():
+            print "attempt #" + str(attempt + 1)
+            print "Knight's tour complete!"
+            print_board(board, rows, cols)
+            sys.exit()
+
+    # if this point is reached (i.e. sys.exit() is not called because a
+    # complete tour was not found) then the knight's tour failed
+    print "knight's tour failed!"
+    print_board(board, rows, cols)
+
+
+def generate_possible_moves(knight_pos, board, rows, cols):
+    """Generate a list of possible moves from a given position."""
+
+    possible_moves = ((-2, -1), (-2, 1), (-1, -2), (-1, 2), (2, -1), (2, 1),
+                      (1, -2), (1, 2))
+    rv = ()
+
+    for (movex, movey) in possible_moves:
+        knight_pos[0] += movex
+        knight_pos[1] += movey
+
+        if (knight_pos[0] >= 0 and
+            knight_pos[1] >= 0 and
+            knight_pos[0] < rows and
+            knight_pos[1] < cols and
+            board[tuple(knight_pos)] == None):
```

Comment:
http://stackoverflow.com/a/3258748

In 99.99% of cases it doesn't matter. It's just a habit to get into for the 0.01% of cases where it might matter. "is None" looks right to experienced python programmers. "== None" looks weird.